sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -45,6 +48,22 @@ static int _decode_value(unsigned val, uint8_t **pkt_pos_ptr, uint8_t *pkt_end);
 static uint32_t _decode_uint(uint8_t *pkt_pos, unsigned nbytes);
 static size_t _encode_uint(uint32_t *val);
 
+__attribute__((section(".noinit")))
+static uint16_t _msg_id;
+uint16_t coap_next_msg_id(void)
+{
+    return atomic_fetch_add_u16(&_msg_id, 1);
+}
+
+static void auto_init_coap_random(void)
+{
+    if (_msg_id == 0) {
```

Comment:
Doesn't this assume that SRAM has a value of `0` on the first boot? To me understanding this is not necessarily the case and this us being relied upon by the PUF modules.

The PUF random seeding utility does track whether it is a cold boot. Maybe we should split this out an own (trivial) module and use that here?

Reply:
I think the standard actually wants the first msg IDs to be randomized. I do agree with `.noinit`, but I doubt that checking it to be zero is a good way to determine it is a cold boot.

Reply:
>With always randomizing there is a small chance of this happening.

especially since random seed might not be random but `RANDOM_SEED_DEFAULT` or `cpuid` in these cases `.noinit` seems a better bet to not reuse the same msg_id.

checking for 0 we catch the cpus that have their memory cells drift to 0 if powered off .

Reply:
> checking for 0 we catch the cpus that have their memory cells drift to 0 if powered off .

Note: Production variance has a significant impact on whether a given SRAM cell tends ti drift to zero or one. A few cells end up being randomly zero or one. With large enough SRAM this is a viable source of entropy.

So, there really is no sense in checking for it being zero. One could profile the memory at the address this gets linked to for a specific SRAM and end up with a classification that would yield a probability for it being a cold boot.

We do have reliable ways to reliably tell the reset cause: Some MCUs readily tell so. For others, we can store a magic cookie value into a `.noinit` variable. If it is still stored there on boot, it was a warm boot.

We may already have a module for that, I think this is ringing a bell.